keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1112 @@
+// PodSpec is a description of a pod.
+type PodSpec struct {
+	...
+	PodGroup PodGroupRef
+	...
+}
+
+// PodGroupRef contains enough information to let you identify an owning
+// pod group.
+// - pod.spec.podGroup.name = foo
+// - pod.spec.podGroup.subset = (bar/baz)
+type PodGroupRef struct {
+	// Name is the name of PodGroup.
+	Name   string
+	// Subset is the name of Subset.
+	Subset string
+}
+```
+
+Pods in the same PodGroup with different priorities might lead to unintended behavior, 
+so need to ensure Pods in the same PodGroup with the same priority.
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators can rely on it to make pending PodGroup schedulable. For this, we 
+need to add a new PodGroup Controller in controller-manager.
+
+```go
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Represents the latest available observations of 
+	// a pod group's current state.
+	Conditions []PodGroupCondition
+
+	// Subsets is a list of SubsetStatus which represents 
+	// the current state of a single subset.
+	Subsets []SubsetStatus
+}
+
+// PodGroupCondition describes the state of a pod group 
+// at a certain point.
+type PodGroupCondition struct {
+	// Type of deployment condition.
+	Type PodGroupConditionType
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	
+	// The reason for the condition's last transition.
+	Reason string
+	
+	// A human readable message indicating details about 
+	// the transition.
+	Message string
+}
+
+type PodGroupConditionType string
+
+const (
+	// PodGroupScheduled represents status of the scheduling 
+	// process for this pod group.
+	PodGroupScheduled PodGroupConditionType = "PodGroupScheduled"
+)
+
+// SubsetStatus represents the current state of a subset.
+type SubsetStatus struct {
+	Name string
+
+	// Total number of non-terminated pods targeted by this subset. (
+	// their labels match the selector)
+	Total int32
+
+	// ScheduledAvailable is the number of pods in the subset that can be scheduled
+	// successfully, but the scheduling fails because minMember cannot be met.
+	ScheduledAvailable int32
+
+	// ScheduledUnavailable is the number of pods in the subset that can't be 
+	// scheduled successfully.
+	ScheduledUnavailable int32
+
+	// Running is the number of pods targeted by this subset
+	// with a Running phase.
+	Running int32
+}
+```
+
+ After the pod group is scheduled for the first time, the condition `PodGroupScheduled` is
+ added by the `pod-group-controller` in kube-controller-manager. Then `pod-group-controller`
+ will reconcile the status of pod group. The workload operator can watch the pod group status,
```

Comment:
What about when we complete the pod group scheduling? We'll update the condition or just delete the podGroup by controller?

Reply:
From below text`The user waits for the PodGroup to be complete and then deletes it.`, I guess the controller will not delete the podGroup automatically, should we add another condition type to indicate we finished the pod group scheduling? Like `Complete` as Job does?

Reply:
Also, the condition for failure.